src/coreclr/jit/decomposelongs.cpp

```diff
@@ -582,44 +588,213 @@ GenTree* DecomposeLongs::DecomposeCast(LIR::Use& use)
+
+            // Now we handle saturation of the result for positive overflow.
+            //
+            // This creates the equivalent of the following C# code:
+            //   var maxFloatingValue = Vector128.Create(9223372036854775808.0);
+            //   var compareMode = FloatComparisonMode.OrderedGreaterThanOrEqualNonSignaling;
+            //   var compareMax = Avx.CompareScalar(srcVec, maxFloatingValue, compareMode);
+
+            NamedIntrinsic compareIntrinsic = (srcType == TYP_FLOAT) ? NI_AVX_Compare : NI_AVX_CompareScalar;
+            GenTreeVecCon* maxFloatingValue = m_compiler->gtNewVconNode(TYP_SIMD16);
+
+            if (srcType == TYP_FLOAT)
+            {
+                // For float->long, we broadcast the comparison value, same as we broadcast the input.
+                for (uint32_t index = 0; index < 4; index++)
+                {
+                    maxFloatingValue->gtSimdVal.f32[index] = 9223372036854775808.0f;
+                }
+            }
+            else
+            {
+                maxFloatingValue->gtSimdVal.f64[0] = 9223372036854775808.0;
+            }
+
+            castRange.InsertAtEnd(maxFloatingValue);
+
+            srcClone             = m_compiler->gtClone(srcVector);
+            GenTree* compareMode = m_compiler->gtNewIconNode(
+                static_cast<int32_t>(FloatComparisonMode::OrderedGreaterThanOrEqualNonSignaling));
+            GenTree* compareMax = m_compiler->gtNewSimdHWIntrinsicNode(TYP_SIMD16, srcClone, maxFloatingValue,
+                                                                       compareMode, compareIntrinsic, srcBaseType, 16);
+
+            castRange.InsertAtEnd(srcClone);
+            castRange.InsertAtEnd(compareMode);
+            castRange.InsertAtEnd(compareMax);
+
+            // We will use the compare mask multiple times, so we replace it with a lclVar.
+            LIR::Use cmpUse;
+            LIR::Use::MakeDummyUse(castRange, compareMax, &cmpUse);
+            cmpUse.ReplaceWithLclVar(m_compiler);
+            compareMax = cmpUse.Def();
+
+            // Mask in long.MaxValue for positive saturation. In the case of overflow, the compare
+            // mask will be all ones. We shift that value right by one to create the MaxValue vector.
+            // This is where we treat two adjacent elements from a float compare as one 64-bit mask.
+            //
+            // This creates the equivalent of the following C# code:
+            //   var maxLong = Sse2.ShiftRightLogical(compareMax, 1);
+            //   castResult = Vector128.ConditionalSelect(compareMax, maxLong, convertResult);
+
+            GenTree* cmpClone = m_compiler->gtClone(compareMax);
+            GenTree* one      = m_compiler->gtNewIconNode(1);
+            GenTree* maxLong  = m_compiler->gtNewSimdHWIntrinsicNode(TYP_SIMD16, compareMax, one,
+                                                                     NI_X86Base_ShiftRightLogical, dstBaseType, 16);
```

Comment:
Does this need to be ShiftRightLogical and such?

Negative values will already saturate correctly (since the sentinel returned is `-9223372036854775808`). So we only need to have fixed up `NaN` (which can be done with `classify` + `embedded mask` on the conversion) or in a post fixup, which can be done via a blend using the `>= MaxValue` mask, the conversion result, and `MaxValue`.

This will end up working better with register selection, especially long term, because it means that we can use the EVEX encoding of the compare instruction and not be forced into the VEX encoded form.